modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,691 @@
+
+template<typename Impl, typename... Ins, typename... Outs>
+struct OAKCallHelper<Impl, std::tuple<Ins...>, std::tuple<Outs...> > {
+    template<int... IIs, int... OIs>
+    static std::shared_ptr<dai::Node> construct_impl(  GOAKContext &ctx
+                                                     , std::vector<std::pair<std::string,
+                                                                             dai::Buffer>>& in_queues_params
+                                                     , cv::detail::Seq<IIs...>
+                                                     , cv::detail::Seq<OIs...>) {
+        return Impl::put(OAKKernelParams{ctx.pipeline(),
+                                         ctx.camera_size(),
+                                         in_queues_params},
+                         get_in<Ins>::get(ctx, IIs)...,
+                         get_out<Outs>::get(ctx, OIs)...);
+    }
+
+    static std::shared_ptr<dai::Node> construct(GOAKContext &ctx,
+                                                std::vector<std::pair<std::string,
+                                                                      dai::Buffer>>& in_queues_params) {
+        return construct_impl(ctx,
+                              in_queues_params,
+                              typename cv::detail::MkSeq<sizeof...(Ins)>::type(),
+                              typename cv::detail::MkSeq<sizeof...(Outs)>::type());
+    }
+};
+
+} // namespace detail
+
+struct GOAKKernel {
+    using F = std::function<std::shared_ptr<dai::Node>(GOAKContext&,
+                                                       std::vector<std::pair<std::string, dai::Buffer>>&)>;
+    explicit GOAKKernel(const F& f) : m_f(f) {}
+    const F m_f;
+};
+
+struct OAKComponent
+{
+    static const char *name() { return "OAK Component"; }
+    GOAKKernel k;
+};
+
+}} // namespace gimpl // namespace cv
+
+using OAKGraph = ade::TypedGraph
+    < cv::gimpl::OAKComponent
+    // FIXME: extend
+    >;
+
+using ConstOAKGraph = ade::ConstTypedGraph
+    < cv::gimpl::OAKComponent
+    // FIXME: extend
+    >;
+
+// This function links dai operation nodes - parent's output to child's input.
+// It utilizes G-API graph to search for operation's node it's previous operation in graph
+// when links them in dai graph.
+void cv::gimpl::GOAKExecutable::LinkToParentHelper(ade::NodeHandle handle,
+                                                   const std::vector<ade::NodeHandle>& nodes)
+{
+    ade::NodeHandle parent;
+    for (const auto& data_nh : handle.get()->inNodes()) {
+        // data_nh - node's input data
+        // need to find which other node produces that data
+        for (const auto& nh : nodes) {
+            if (m_gm.metadata(nh).get<NodeType>().t == NodeType::OP) {
+                for (const auto& outdatah : nh.get()->outNodes()) {
+                    if (data_nh == outdatah) {
+                        parent = nh;
+                    }
+                }
+            }
+        }
+    }
```

Comment:
There is only one producer for each data node, so you simply can get inNodes().front() for each input data node without iterating on every island node

Reply:
Reworked